test/setter-tests.js

```diff
@@ -57,4 +57,41 @@ describe("parent setters", () => {
     assert.strictEqual(secondCallCount, 2);
     assert.strictEqual(value, "second:7");
   });
+
+  it("should only be discarded if getter did not throw", () => {
+    // Manually register setters for value and set, which are marked as
+    // constant in the fake-const.js module, even though they are actually
+    // mutable (see below).
+    let value, set;
+    let valueSetterCallCount = 0;
+    module.watch(require("./setter/fake-const.js"), {
+      value: v => {
+        value = v;
+        valueSetterCallCount += 1;
+      },
+      set: v => set = v
+    });
+
+    // Because the getter for value in fake-const.js throws, the initial
+    // value here is undefined, and our live binding to that value is
+    // maintained for now.
+    assert.strictEqual(value, void 0);
+    assert.strictEqual(valueSetterCallCount, 1);
+
+    // Since the exported value is marked as constant, the setter that
+    // updates our copy of value should be discarded as soon as it has
+    // been called with the final value, but this set(2) still works
+    // because the getter threw the last time it was called. After this,
+    // however, the setter function will be discarded.
```

Comment:
Although this whole test is a bit artificial, this logic breaks if we remove the `getter.runCount > 0` test in `lib/entry.js`, because we forget the setter too soon.